### PR TITLE
refactor: update CollisionDetection interface to receive active draggable node, collision rect and array of droppable containers

### PR DESCRIPTION
**Pull request by wmain**
Resolves https://github.com/clauderic/dnd-kit/issues/255 and https://github.com/clauderic/dnd-kit/issues/309.

This updates the signature of CollisionDetection to accept `DraggingNode` and `DroppableContainer[]` in its arguments, hopefully making it easier to build richer custom collision detection functions. I've also made a small performance improvement to the three included collision detection algorithms, which were making an unnecessary second pass to determine the index of the relevant figure.

The one uncertainty I had writing this was the inclusion of the new `DraggingNode` type. I chose an additional type because `rect` for `DraggableNode` would always be null for the non active node, and this would lead to a bunch of unnecessary non-null assertions in collision detection algorithms. 

Looking forward to feedback, and thanks for the great library!

### CODE REVIEW

**Comment by wmain** [comment]
@clauderic Unsure if you'd prefer to write the changeset or if I should include that as well. The bot doesn't really suggest which is better :).

**Comment by wmain** [comment]
Looks like some unrelated changes were picked up by `lint --fix`. Happy to remove if you'd like.

**Comment by clauderic** [comment]
Hey @wmain, thanks for getting the ball rolling on this one and opening a PR 🙏  I've been swamped recently and haven't had time to review this yet, but should be able to get to it next week.

**Comment by wmain** [comment]
Hey @clauderic, anything I can do to help with this? Trying not to rush you, but this is blocking some work we want to do on our project.

**Comment by clauderic** [comment]
Just an update, I'm back from vacation this week, will get this merged before end of week

File: packages/core/src/components/DndContext/DndContext.tsx [hunk 1]
```diff
@@ -255,9 +256,27 @@ export const DndContext = memo(function DndContext({
...
+    if (!activeNode) return null;
+
+    return {...activeNode, rect: collisionRect};
+  }, [collisionRect, draggableNodes]);
+
+  const definedDroppableContainers = useMemo(() => {
+    return Object.values(droppableContainers).filter(
+      Boolean
+    ) as DroppableContainer[];
+  }, [droppableContainers]);
```
Comment (clauderic) [inline packages/core/src/components/DndContext/DndContext.tsx:272]: Could you give me a bit of context on these changes @wmain?

Comment (wmain) [reply packages/core/src/components/DndContext/DndContext.tsx:272]: Per our discussion in #255, lines 259-266 memoize the dragging node to be passed into the provided collision detection function.

`definedDroppableContainers` seeks to provide only the defined droppable containers to the collision detection function, as [`DroppableContainers`](https://github.com/clauderic/dnd-kit/blob/54c877875cf7ec6d4367ca11ce216cc3eb6475d2/packages/core/src/store/types.ts#L59) notes that for any one record the droppable container might be undefined.

**Review by clauderic — APPROVED** [review]
Thanks for getting the ball rolling on this one @wmain, should be good to go 🚢 

I won't be releasing this to npm on the `latest` tag just yet, but you can start using it by installing the `next` tag:

```bash
npm install @dnd-kit/core@next @dnd-kit/sortable@next
```

**Comment by wmain** [comment]
Thanks @clauderic !

**Comment by tains2384** [comment]
> I won't be releasing this to npm on the `latest` tag just yet, but you can start using it by installing the `next` tag:

Hi @clauderic ,

I just follow example on storybook (Preset/Sortable/Multiple Containers/Basic Setup) and it takes me lot of time to find out this change not release on the latest tag. Please note this in document or storybook incase anyone want to refer

BTW, thank you for your work.

**Pull request merged**